server/instances.go

```diff
@@ -151,6 +151,29 @@ func (p *Plugin) InstallInstance(instance Instance) error {
 				}
 			}
 
+			if instance.Common().Type == CloudOAuthInstanceType && len(instances.IDs()) > 0 && instances.checkIfExists(instance.GetID()) {
+				p.API.LogDebug("Getting the stored JWT instance data to store it inside the OAuth instance data.")
+				jwtInstance, err := p.instanceStore.LoadInstance(instance.GetID())
+				if err != nil {
+					p.API.LogError("Error occurred while fetching the instance", "ID", instance.GetID(), "Error", err.Error())
+					return err
+				}
+
+				// Storing the JWT instance data inside the OAuth instance data. We will use this to use the stored JWT token in case the user has not connected to OAuth yet.
+				p.API.LogDebug("Instance type stored in KV store", "ID", jwtInstance.GetID(), "Type", jwtInstance.Common().Type)
+				if jwtInstance.Common().Type == CloudInstanceType {
+					instance.(*cloudOAuthInstance).JWTInstance = jwtInstance.(*cloudInstance)
+				} else if jwtInstance.Common().Type == CloudOAuthInstanceType {
+					instance.(*cloudOAuthInstance).JWTInstance = jwtInstance.(*cloudOAuthInstance).JWTInstance
```

Comment:
This seems a bit weird to me but also makes sense. Is the point of this line is to make sure we handle the case where we run `/jira instance install cloud-oauth` twice? We "carry over" the JWT instance embedded inside the existing one. This seems a bit error-prone since this is the only field we're carrying over. Are there any other fields we want to retain when "over-installing" a `cloud-oauth` over an existing `cloud-oauth` instance?

Maybe put a comment here stating this is here to make sure we carry it over when we install `cloud-oauth` twice

Reply:
It seems like a null pointer is not possible here but still a little concerned about it. Ideally we always do the `ok` syntax to avoid type assertion panics

Reply:
@mickmister This function "installInstance" is being called from the "installCloudOAuthInstance" function. After running the install command "installCloudOAuthInstance" is called twice during the flow:
1. When we run the install command.
2. When we submit the clientID and clientSecret modal. (To store the credentials in the instance data)

The instance is stored in the KV store after running the install command and the JWT instance data is stored inside it (Now the previous JWT instance data is overridden by the oauth instance data). When the function is called on submitting the modal, if we don't carry forward the JWT instance data then it will be lost.

Reply:
> When the function is called on submitting the modal, if we don't carry forward the JWT instance data then it will be lost.

Yeah when reviewing the OAuth implementation I thought it was weird how we call "install" when submitting the credential form. Being able to run `/jira install` on an existing instance was also introduced as a result, which also seemed off.

The fact that we have to manually add the `JWTInstance` here without making sure we copy everything over is a little concerning. If we add more fields to the `cloud-oauth` instance, we have to update this piece of code, and that seems error-prone.

Also I wonder if there are unwanted side effects to running `/jira install server myurl` after an existing Jira Server has been installed with that same URL. We may potentially lose information on the existing struct.

---

I think we can make the code more defensive by changing the "clientSecret modal" logic to be:

```go
if existingInstance != nil {
    existingInstance.ClientID = clientID
    existingInstance.ClientSecret = clientSecret
    storeInstance(existingInstance)
} else {
    newInstance := makeNewInstance()
    installInstance(newInstance)
}
```

And then disallow calling `installInstance` with an existing instance, just as it was before the OAuth code was introduced. `/jira install` can just start a setup flow if an instance exists.

Reply:
@raghavaggarwal2308 What do you think about my comment above?

Reply:
@mickmister I agree with your suggestion. I have made the necessary changes according to the same.

Note: I have updated the condition to override the enterprise check to work only when a user is migrating from JWT to OAuth. Now if we try to run the "/jira install cloud-oauth ..." OR "/jira install server ..." command twice then we will get that message "You need a valid enterprise licence ....".